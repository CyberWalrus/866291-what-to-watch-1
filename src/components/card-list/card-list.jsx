import React from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import Card from "../card/card.jsx";

class CardList extends React.Component {
  constructor(props) {
    super(props);
  }
  componentDidMount() {}
  onCardMouseEnter() {}
  render() {
    return (
      <div className="catalog__movies-list">
        {this.props.films &&
          this.props.films.map(({id, title, src, genre, preview}) => (
            <Card
              key={id}
              id={id}
              title={title}
              src={src}
              genre={genre}
              preview={preview}
              onMouseEnter={this.onCardMouseEnter}
            />
          ))}
      </div>
    );
  }
}

CardList.propTypes = {
  films: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        src: PropTypes.string.isRequired,
        genre: PropTypes.oneOf([
          `comedy`,
          `crime`,
          `documentary`,
          `drama`,
          `horror`,
          `family`,
          `romance`,
          `sci-fi`,
          `thriller`
        ]).isRequired,
        preview: PropTypes.string.isRequired
      })
  )
};
const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    films: state.films
  });

export {CardList};

export default connect(mapStateToProps)(CardList);
