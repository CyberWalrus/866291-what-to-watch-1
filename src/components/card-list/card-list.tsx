import * as React from "react";
import {Fragment} from "react";
import {connect} from "react-redux";
import Card from "../card/card";
import {getFilms, getFavorites} from "../../store/data/selectors";
import {Film} from "../../type/data";
import {StateApp} from "../../type/reducer";

interface PropsInsert {
  filmId: number
  numberFilm: number,
  activeFilm: number,
  genreFilm: string,
  isFavorite: boolean,
  onClickShowMore: () => void,
  onMouseEnterCard: (id: number) => void,
  onMouseLeaveCard: (id: number) => void,
  onClickToRedirect: (id: number) => void
}
interface PropsState {
  films: Film[]
}

type Props = PropsInsert & PropsState;

const CardList = ({
  films,
  numberFilm,
  activeFilm,
  onClickShowMore,
  onMouseEnterCard,
  onMouseLeaveCard,
  onClickToRedirect,
  genreFilm
}: Props) => {
  const isShow = genreFilm ? false : true;
  return (
    <Fragment>
      <div className="catalog__movies-list">
        {films &&
          films.map(({id, title, srcPreviewImage, srcPreviewVideo}) => (
            <Card
              key={id}
              id={id}
              title={title}
              srcPreviewImage={srcPreviewImage}
              srcPreviewVideo={srcPreviewVideo}
              isActive={activeFilm === id}
              onMouseEnterCard={() => onMouseEnterCard(id)}
              onMouseLeaveCard={() => onMouseLeaveCard(id)}
              onClickToRedirect={() => onClickToRedirect(id)}
            />
          ))}
      </div>
      {films && isShow && films.length >= numberFilm ? (
        <div className="catalog__more">
          <button
            className="catalog__button"
            type="button"
            onClick={onClickShowMore}
          >
            Show more
          </button>
        </div>
      ) : (
        <Fragment />
      )}
    </Fragment>
  );
};

const mapStateToProps = (state: StateApp, ownProps: Props): Props =>
  Object.assign({}, ownProps, {
    films: ownProps.isFavorite
      ? getFavorites(state, ownProps.numberFilm)
      : getFilms(
        state,
        ownProps.numberFilm,
        ownProps.genreFilm,
        ownProps.filmId
      )
  });

export {CardList};

export default connect(mapStateToProps)(CardList);
