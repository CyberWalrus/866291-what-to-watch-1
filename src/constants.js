const GENRE_DEFOULT = `All genres`;
const NUMBER_FILM = 20;
const NUMBER_FILM_GENRE = 4;
const SERVER_URL = `https://es31-server.appspot.com/wtw`;
const SERVER_URL_USER = `https://es31-server.appspot.com/`;
const LINK_STYLE = `inherit`;
const MAIN_FILM_ID = 1;
const PADDING_VIDEO = 25;
const REVIEW_MESSAGE = `OK`;

const ratingRadioValues = [`1`, `2`, `3`, `4`, `5`];

const RatingName = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`
};
const RatingInterval = {
  BAD: 3,
  NORMAL: 5,
  GOOD: 8,
  VERY_GOOD: 10,
  AWESOME: 10
};
const FilmRoute = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};
const BodyOverflow = {
  VISIBLE: `visible`,
  HIDDEN: `hidden`
};
const FavoriteStatus = {
  ADD: 1,
  REMOVE: 0
};
const TextLength = {
  MIN: 50,
  MAX: 400
};
const RatingValue = {
  MIN: 1,
  MAX: 5
};

const DisabledStyle = {
  opacity: `0.4`,
  border: `inherit`,
  cursor: `not-allowed`
};
const OptionsVideoMin = {
  width: `280`,
  height: `175`,
  isMuted: true,
  isLoop: true,
  isControls: false
};
const OptionsVideoFull = {
  width: `100%`,
  height: `100%`,
  isMuted: false,
  isLoop: false,
  isControls: false
};

export {
  GENRE_DEFOULT,
  NUMBER_FILM,
  SERVER_URL,
  SERVER_URL_USER,
  NUMBER_FILM_GENRE,
  LINK_STYLE,
  PADDING_VIDEO,
  REVIEW_MESSAGE,
  MAIN_FILM_ID,
  ratingRadioValues,
  RatingName,
  RatingInterval,
  FilmRoute,
  OptionsVideoMin,
  OptionsVideoFull,
  BodyOverflow,
  FavoriteStatus,
  DisabledStyle,
  TextLength,
  RatingValue
};
