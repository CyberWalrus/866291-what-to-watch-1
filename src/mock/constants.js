const FilmRoute = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};
const GENRE_DEFOULT = `All genres`;
const NUMBER_FILM = 20;

export {FilmRoute, GENRE_DEFOULT, NUMBER_FILM};

