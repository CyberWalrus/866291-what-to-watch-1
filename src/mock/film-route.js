const FilmRoute = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};
export default FilmRoute;
