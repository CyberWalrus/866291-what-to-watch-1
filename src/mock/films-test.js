const films = [
  {
    id: 1,
    genre: `drama`,
    title: `Aviator`,
    src: `img/aviator.jpg`
  },
  {
    id: 2,
    genre: `drama`,
    title: `Macbeth`,
    src: `img/macbeth.jpg`
  }
];
export default films;
