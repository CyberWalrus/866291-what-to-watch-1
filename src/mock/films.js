const films = [
  {
    id: 1,
    genre: `drama`,
    title: `Aviator`,
    src: `img/aviator.jpg`
  },
  {
    id: 2,
    genre: `drama`,
    title: `Macbeth`,
    src: `img/macbeth.jpg`
  },
  {
    id: 3,
    genre: `drama`,
    title: `Revenant`,
    src: `img/revenant.jpg`
  },
  {
    id: 4,
    genre: `comedy`,
    title: `Johnny English`,
    src: `img/johnny-english.jpg`
  },
  {
    id: 5,
    genre: `comedy`,
    title: `What We Do in the Shadows`,
    src: `img/what-we-do-in-the-shadows.jpg`
  }
];
export default films;
