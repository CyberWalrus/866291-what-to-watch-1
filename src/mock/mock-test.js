const FILMS = [
  {
    bgColor: `#977461`,
    description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
    director: `Martin Scorsese`,
    genre: `Thriller`,
    id: 1,
    isFavorite: false,
    pageUrl: `Shutter Island`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    rating: `8,1`,
    ratingLevel: `Very good`,
    released: 2010,
    runTime: 138,
    scoresCount: 1002557,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
    starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
    title: `Shutter Island`
  },
  {
    bgColor: `#F1E9CE`,
    description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
    director: `Justin Kurzel`,
    genre: `Drama`,
    id: 2,
    isFavorite: false,
    pageUrl: `Macbeth`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: `6,6`,
    ratingLevel: `Good`,
    released: 2015,
    runTime: 113,
    scoresCount: 48798,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg`,
    starrings: [(`Michael Fassbender`, `Marion Cotillard`, `Jack Madigan`)],
    title: `Macbeth`
  }
];
const FILM = {
  bgColor: `#977461`,
  description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
  director: `Martin Scorsese`,
  genre: `Thriller`,
  id: 1,
  isFavorite: false,
  pageUrl: `Shutter Island`,
  srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  rating: `8,1`,
  ratingLevel: `Very good`,
  released: 2010,
  runTime: 138,
  scoresCount: 1002557,
  srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
  srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
  srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
  starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
  title: `Shutter Island`
};
const REVIEWS = [
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  },
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  }
];

const GENRES = [
  `All genres`,
  `Thriller`,
  `Drama`,
  `Adventure`,
  `Comedy`,
  `Crime`,
  `Action`,
  `Fantasy`
];
export {FILMS, FILM, GENRES, REVIEWS};
