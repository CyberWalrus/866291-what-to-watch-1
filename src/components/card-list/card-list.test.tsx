import * as React from "react";
import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import {shallow} from "enzyme";
import {BrowserRouter} from "react-router-dom";
import {CardList} from "./card-list";
import {FILMS} from "../../mock/data-mock";
import {NUMBER_FILM, GENRE_DEFOULT} from "../../constants";

Enzyme.configure({adapter: new Adapter()});
it(`CardList component renders correctly with props`, (): void => {
  const handleClick = jest.fn();
  const tree = toJson(
    shallow(
      <BrowserRouter>
        <CardList
          filmId={1}
          films={FILMS}
          numberFilm={NUMBER_FILM}
          activeFilm={0}
          isFavorite={true}
          onClickShowMore={handleClick}
          onMouseEnterCard={handleClick}
          onMouseLeaveCard={handleClick}
          onClickToRedirect={handleClick}
          genreFilm={GENRE_DEFOULT}
        />
      </BrowserRouter>
    )
  );
  expect(tree).toMatchSnapshot();
});
it(`CardList component renders correctly with empty props`, (): void => {
  const handleClick = jest.fn();
  const tree = toJson(
    shallow(
      <BrowserRouter>
        <CardList
          filmId={1}
          films={[]}
          numberFilm={NUMBER_FILM}
          activeFilm={0}
          isFavorite={true}
          onClickShowMore={handleClick}
          onMouseEnterCard={handleClick}
          onMouseLeaveCard={handleClick}
          onClickToRedirect={handleClick}
          genreFilm={GENRE_DEFOULT}
        />
      </BrowserRouter>
    )
  );
  expect(tree).toMatchSnapshot();
});
