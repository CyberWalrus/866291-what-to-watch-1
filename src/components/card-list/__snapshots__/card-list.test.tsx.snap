// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList component renders correctly with empty props 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <CardList
    activeFilm={0}
    filmId={1}
    films={Array []}
    genreFilm="All genres"
    isFavorite={true}
    numberFilm={20}
    onClickShowMore={[MockFunction]}
    onClickToRedirect={[MockFunction]}
    onMouseEnterCard={[MockFunction]}
    onMouseLeaveCard={[MockFunction]}
  />
</Router>
`;

exports[`CardList component renders correctly with props 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <CardList
    activeFilm={0}
    filmId={1}
    films={
      Array [
        Object {
          "bgColor": "#977461",
          "description": "In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.",
          "director": "Martin Scorsese",
          "genre": "Thriller",
          "id": 1,
          "isFavorite": false,
          "pageUrl": "Shutter Island",
          "rating": "8,1",
          "ratingLevel": "Very good",
          "released": 2010,
          "runTime": "1h 44m",
          "scoresCount": 1002557,
          "srcBgImage": "https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg",
          "srcPosterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg",
          "srcPreviewImage": "https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg",
          "srcPreviewVideo": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          "srcVideo": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
          "starrings": Array [
            "Leonardo DiCaprio",
            "Emily Mortimer",
            "Mark Ruffalo",
          ],
          "title": "Shutter Island",
        },
        Object {
          "bgColor": "#F1E9CE",
          "description": "Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.",
          "director": "Justin Kurzel",
          "genre": "Drama",
          "id": 2,
          "isFavorite": false,
          "pageUrl": "Macbeth",
          "rating": "6,6",
          "ratingLevel": "Good",
          "released": 2015,
          "runTime": "1h 44m",
          "scoresCount": 48798,
          "srcBgImage": "https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg",
          "srcPosterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg",
          "srcPreviewImage": "https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg",
          "srcPreviewVideo": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
          "srcVideo": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
          "starrings": Array [
            "Michael Fassbender",
            "Marion Cotillard",
            "Jack Madigan",
          ],
          "title": "Macbeth",
        },
      ]
    }
    genreFilm="All genres"
    isFavorite={true}
    numberFilm={20}
    onClickShowMore={[MockFunction]}
    onClickToRedirect={[MockFunction]}
    onMouseEnterCard={[MockFunction]}
    onMouseLeaveCard={[MockFunction]}
  />
</Router>
`;
